nikola/plugin_categories.py

```diff
@@ -372,6 +379,71 @@ class MarkdownExtension(CompilerExtension):
     compiler_name = "markdown"
 
 
+class MetadataExtractor(BasePlugin):
+    """Plugins that can extract meta information from post files."""
+
+    # Name of the extractor. (required)
+    name = "unknown"
+    # Where to get metadata from. (MetaSource; required)
+    source = None
+    # Priority of extractor. (MetaPriority; required)
+    priority = None
+    # List of tuples (MetaCondition, arg) with conditions used to select this extractor.
+    conditions = []
+    # Regular expression used for splitting metadata, or None if not applicable.
+    split_metadata_re = None
+    # List of tuples (import name, pip name, friendly name) of Python packages required for this extractor.
+    requirements = []
+    # Name of METADATA_MAPPING to use, if any.
+    map_from = None
+    # Whether or not the extractor supports writing metadata.
+    supports_write = False
+
+    def _extract_metadata_from_text(self, source_text: str) -> dict:
+        """Extract metadata from text."""
+        raise NotImplementedError()
+
+    def split_metadata_from_text(self, source_text: str) -> (str, str):
+        """Split text into metadata and content (both strings)."""
+        if self.split_metadata_re is None:
+            return source_text
+        else:
+            split_result = self.split_metadata_re.split(source_text.lstrip(), maxsplit=1)
+            if len(split_result) == 1:
+                return split_result[0], split_result[0]
```

Comment:
You can’t know what is metadata and what is text if splitting fails.

The old implementation did it differently: it always split text into lines, and kept looking for data until it found an empty line. YAML/TOML was hacked onto it — if it found the first line, it would have to re-join text after finding the second delimiter. There were two different splitting algorithms, one based on `str.split('\n')` and the other on regexes.

This implementation unifies splitting with the regex-based solution, and `str.split('\n')` happens only in the Nikola extractor. It also allows YAML/TOML in 2-file posts. But there’s a side effect of that: since we always call `split_metadata_from_text`, even for 2-file posts, it needs to return the same string twice if it doesn’t find the delimiter.

I consider this to be good-enough, because we always use only one of this function’s return values.